File: packages/core-py/src/mcp_relay_core/oauth/jwt_issuer.py
"""RSA JWT Issuer and JWKS generation helper."""

import datetime
from pathlib import Path
from typing import cast

import jwt
from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric import rsa

# Keys will be stored outside of the codebase to persist across server restarts
DEFAULT_KEYS_DIR = Path.home() / ".mcp-relay" / "jwt-keys"


class JWTIssuer:
    def __init__(self, server_name: str, keys_dir: Path = DEFAULT_KEYS_DIR):
        self.server_name = server_name
        self.keys_dir = keys_dir
        self.private_key_path = self.keys_dir / f"{server_name}_private.pem"
        self.public_key_path = self.keys_dir / f"{server_name}_public.pem"

        self.private_key: rsa.RSAPrivateKey | None = None
        self.public_key: rsa.RSAPublicKey | None = None
        self._kid = "key-1"
        self._load_or_generate_keys()

    def _load_or_generate_keys(self) -> None:
        self.keys_dir.mkdir(parents=True, exist_ok=True)

        if self.private_key_path.exists() and self.public_key_path.exists():
            with open(self.private_key_path, "rb") as f:
                self.private_key = cast(
                    rsa.RSAPrivateKey,
                    serialization.load_pem_private_key(f.read(), password=None),
                )
            with open(self.public_key_path, "rb") as f:
                self.public_key = cast(
                    rsa.RSAPublicKey, serialization.load_pem_public_key(f.read())
                )
        else:
            self.private_key = rsa.generate_private_key(
                public_exponent=65537,
                key_size=2048,
            )
            self.public_key = self.private_key.public_key()

            with open(self.private_key_path, "wb") as f:
                f.write(
                    self.private_key.private_bytes(
                        encoding=serialization.Encoding.PEM,
                        format=serialization.PrivateFormat.PKCS8,
                        encryption_algorithm=serialization.NoEncryption(),
                    )
                )

            with open(self.public_key_path, "wb") as f:
                f.write(
                    self.public_key.public_bytes(
                        encoding=serialization.Encoding.PEM,
                        format=serialization.PublicFormat.SubjectPublicKeyInfo,
                    )
                )
            # Ensure proper file permissions
            self.private_key_path.chmod(0o600)
            self.public_key_path.chmod(0o644)

    def get_jwks(self) -> dict:
        """Return JWKS payload for /.well-known/jwks.json"""
        if self.public_key is None:
            raise RuntimeError("Public key not loaded")

        pn = self.public_key.public_numbers()

        def to_base64url(val: int) -> str:
            val_bytes = val.to_bytes((val.bit_length() + 7) // 8, byteorder="big")
            # Custom base64url without padding
            import base64

            return base64.urlsafe_b64encode(val_bytes).rstrip(b"=").decode("ascii")

        return {
            "keys": [
                {
                    "kty": "RSA",
                    "use": "sig",
                    "alg": "RS256",
                    "kid": self._kid,
                    "n": to_base64url(pn.n),
                    "e": to_base64url(pn.e),
                }
            ]
        }

    def issue_access_token(self, sub: str, expires_in_seconds: int = 3600) -> str:
        """Issue an RS256 JWT access token."""
        if self.private_key is None:
            raise RuntimeError("Private key not loaded")

        now = datetime.datetime.now(datetime.UTC)
        payload = {
            "iss": self.server_name,
            "aud": self.server_name,
            "sub": sub,
            "iat": now,
            "exp": now + datetime.timedelta(seconds=expires_in_seconds),
        }
        return jwt.encode(
            payload, self.private_key, algorithm="RS256", headers={"kid": self._kid}
        )

    def verify_access_token(self, token: str) -> dict:
        """Verify JWT and return payload. Raises standard PyJWT exceptions on failure."""
        if self.public_key is None:
            raise RuntimeError("Public key not loaded")

        return jwt.decode(
            token,
            self.public_key,
            algorithms=["RS256"],
            audience=self.server_name,
            issuer=self.server_name,
        )

File: packages/core-ts/src/oauth/jwt-issuer.ts
/**
 * RSA JWT Issuer and JWKS generation for TypeScript MCP servers.
 * Uses jose (industry-standard, zero-dep, Web Crypto compatible).
 */

import { existsSync, mkdirSync, readFileSync, writeFileSync } from 'node:fs'
import { homedir } from 'node:os'
import { join } from 'node:path'
import * as jose from 'jose'

const DEFAULT_KEYS_DIR = join(homedir(), '.mcp-relay', 'jwt-keys')

export class JWTIssuer {
  private serverName: string
  private keysDir: string
  private privateKeyPath: string
  private publicKeyPath: string
  private kid = 'key-1'
  // biome-ignore lint/suspicious/noExplicitAny: jose does not export KeyLike in all environments
  private privateKey: any | null = null
  // biome-ignore lint/suspicious/noExplicitAny: jose does not export KeyLike in all environments
  private publicKey: any | null = null
  private _initialized = false

  constructor(serverName: string, keysDir = DEFAULT_KEYS_DIR) {
    this.serverName = serverName
    this.keysDir = keysDir
    this.privateKeyPath = join(this.keysDir, `${serverName}_private.pem`)
    this.publicKeyPath = join(this.keysDir, `${serverName}_public.pem`)
  }

  /** Must call before using issuer — loads or generates RSA keys. */
  async init(): Promise<void> {
    if (this._initialized) return
    mkdirSync(this.keysDir, { recursive: true })

    if (existsSync(this.privateKeyPath) && existsSync(this.publicKeyPath)) {
      const privatePem = readFileSync(this.privateKeyPath, 'utf-8')
      const publicPem = readFileSync(this.publicKeyPath, 'utf-8')
      this.privateKey = await jose.importPKCS8(privatePem, 'RS256')
      this.publicKey = await jose.importSPKI(publicPem, 'RS256')
    } else {
      const { publicKey, privateKey } = await jose.generateKeyPair('RS256', {
        modulusLength: 2048,
        extractable: true
      })
      this.privateKey = privateKey
      this.publicKey = publicKey

      const privatePem = await jose.exportPKCS8(privateKey)
      const publicPem = await jose.exportSPKI(publicKey)

      writeFileSync(this.privateKeyPath, privatePem, { mode: 0o600 })
      writeFileSync(this.publicKeyPath, publicPem, { mode: 0o644 })
    }
    this._initialized = true
  }

  /** Return JWKS payload for /.well-known/jwks.json */
  async getJwks(): Promise<jose.JSONWebKeySet> {
    if (!this.publicKey) throw new Error('JWTIssuer not initialized')
    const jwk = await jose.exportJWK(this.publicKey)
    jwk.kid = this.kid
    jwk.use = 'sig'
    jwk.alg = 'RS256'
    return { keys: [jwk] }
  }

  /** Issue an RS256 JWT access token. */
  async issueAccessToken(sub: string, expiresInSeconds = 3600): Promise<string> {
    if (!this.privateKey) throw new Error('JWTIssuer not initialized')
    return new jose.SignJWT({ sub })
      .setProtectedHeader({ alg: 'RS256', kid: this.kid })
      .setIssuer(this.serverName)
      .setAudience(this.serverName)
      .setIssuedAt()
      .setExpirationTime(`${expiresInSeconds}s`)
      .sign(this.privateKey)
  }

  /** Verify JWT and return payload. Throws on failure. */
  async verifyAccessToken(token: string): Promise<jose.JWTPayload> {
    if (!this.publicKey) throw new Error('JWTIssuer not initialized')
    const { payload } = await jose.jwtVerify(token, this.publicKey, {
      issuer: this.serverName,
      audience: this.serverName
    })
    return payload
  }
}
